src/openforms/submissions/exports.py

```diff
@@ -43,32 +44,54 @@ def create_submission_export(queryset: SubmissionQuerySet) -> tablib.Dataset:
     """
     Turn a submissions queryset into a tablib dataset for export.
 
-    .. note:: the queryset of submissions must all be of the same form!
+    Note each submission can potentially have different fields and variables,
+      so we'll have to do some work to get a dataset in the right shape.
+
+    .. note:: the queryset of submissions technically doesn't have to be from the same Form (anymore)
     """
     # queryset *could* be empty
-    if not queryset:
+    if not queryset.exists():
         return tablib.Dataset()
 
-    first_submission = queryset[0]
-    headers = ["Formuliernaam", "Inzendingdatum"]
-    for data_node in iter_submission_data_nodes(first_submission):
-        if hasattr(data_node, "component"):
-            headers.append(data_node.component["key"])
-        elif hasattr(data_node, "variable"):
-            headers.append(data_node.variable.key)
+    fixed_headers = ["Formuliernaam", "Inzendingdatum"]
+
+    # let's collect unique headers and rows with actual values from all our submissions
+    seen_keys = OrderedSet()
+    rows = list()
+
+    for submission in queryset.order_by("form", "completed_on", "id"):
+        row = dict()
+        for data_node in iter_submission_data_nodes(submission):
+            key = None
+
+            # TODO move key grab to a property on the Node types?
+            if hasattr(data_node, "component"):
+                key = data_node.component["key"]
+            elif hasattr(data_node, "variable"):
+                key = data_node.variable.key
 
+            if key:
+                seen_keys.add(key)
+                row[key] = data_node.value
+
+        rows.append((submission, row))
+
+    # output table of everything we have
+    headers = fixed_headers + list(seen_keys)
     data = tablib.Dataset(headers=headers)
 
-    for submission in queryset:
+    for submission, row in rows:
         inzending_datum = (
             make_naive(submission.completed_on) if submission.completed_on else None
         )
         submission_data = [
             submission.form.admin_name,
             inzending_datum,
-            *[data_node.value for data_node in iter_submission_data_nodes(submission)],
+            # use blank for missing values
+            *(row.get(key, "") for key in seen_keys),
         ]
         data.append(submission_data)
+
```

Comment:
While typing this out, I'm also wondering if the problem is not "simpler" that there's a bug with hidden (layout) components not showing for `export` RenderMode in `iter_submission_data_nodes` - as the initial bug report was about different submissions for the same form, so they should all have the same keys and all keys in the form (visible or hidden) should be exported always, see https://github.com/open-formulieren/open-forms/blob/master/src/openforms/formio/rendering/conf.py#L14 Perhaps the approach shouldn't be to grab the keys/headers from the first submission, but rather from the entire form of the (first) submission.

---

Thoughts before I realized the above paragraph

In general this approach looks good to me, I'm a bit concerned about memory usage for large datasets though - we're storing all the data of the entire queryset in memory in the `rows` variable before even initializing a tablib dataset, in addition to already having the entire result set loaded in the submission queryset cache. Tablib in itself already isn't the greatest for memory usage (all data is loaded into memory rather than being able to stream/chunk it), so paying this penalty twice/thrice seems like something to avoid - especially if we need to replace tablib in the future.

With that context in mind...

* Can we leverage `queryset.iterator()` so we get memory-efficient (and server-side cursors!) processing of the submissions?
* Perhaps we could set up the tablib dataset all the way at the beginning, and mutate `dataset.headers` on the fly when we detect a new key?
* Then every iteration of the loop over the `queryset.iterator` can add the tablib data row with the correct dimensions (after the first iteration, you should only have to add 'padding' on the right to match the length of `dataset.headers`, as this will never shrink)